worker_tasks/categorize.py

```diff
@@ -46,109 +79,374 @@
     "with", "within", "without", "would", "yet", "you", "your", "yours", "yourself", \
     "yourselves" ] # Source: http://xpo6.com/list-of-english-stop-words/
 
+
 class Categorizer:
 
     def __init__(self):
-        self.title_categories = Counter()
+        self._memoized_related_articles = defaultdict(list)
+        self._memoized_article_links = defaultdict(list)
+        self._memoized_semantic_relatedness_scores = defaultdict(float)
+
 
     def categorize_item(self, title, text):
-        categories = Counter()
+        # Use old word count approach to get possible keywords
+        keyword_candidates = self._get_keyword_candidates(title, text)
+        
+        # Store possible wiki articles and their links
+        self._memoize_related_articles(keyword_candidates.keys())
+        self._memoize_article_links(keyword_candidates.keys())
 
-        # Extract words present in category dictonary
-        title = filter(self.is_from_titles, self.get_valid_ngrams(title))
-        text = filter(self.is_from_titles, self.get_valid_ngrams(text))
+        # Assign best article for each possible keyword 
+        assigned_articles   = self._assign_closest_articles(keyword_candidates.keys())
+        
+        # Build a graph and identify keyword clusters
+        semantic_graph      = self._build_semantic_graph(assigned_articles)
+        clustering          = semantic_graph.community_multilevel(weights='weight')
+        
+        # Use keywords from best clusters
+        selected_categories = self._get_best_keywords(clustering, keyword_candidates)
 
-        # Get word counts from title, text of article
-        categories.update(title + text)
+        return selected_categories
 
-        categories = self.get_weighted_categories(categories)
 
-        return self.get_best_categories(categories)
+    def _get_keyword_candidates(self, title, text):
+        title_ngrams = self._get_valid_ngrams(title)
+        text_ngrams = self._get_valid_ngrams(text)
 
+        ngrams = title_ngrams + text_ngrams
 
-    def reset_title_categories(self):
-        self.title_categories.clear()
+        # Apply weights...
+        for ngram, count in ngrams.iteritems():
+            
+            # ...from title (x2)
+            if ngram in title_ngrams:
+                ngrams.update({ngram: count})
 
-        one_day_ago = datetime.now() - timedelta(hours=24)
-        titles = [item.title for item in Item.query.filter(Item.category_1==None, Item.last_updated>=one_day_ago).all()]
-        self.parse_title_categories(titles)
+            # ...bigram (x2) or trigram (x4)
+            for n in range(ngram.count(' ')):
```

Comment:
That wouldn't work since ngram is just a string. Length of a string would be the number of characters. We could do a split to break the ngram into an array of the grams, but that seemed unnecessary.